script.module.web-pdb/libs/web_pdb/asyncore_wsgi/__init__.py

```diff
@@ -0,0 +1,280 @@
+        try:
+            handler.run(self.server.get_app())
+        except Exception:
+            self.handle_error()
+            return
+        if self.close_connection:
+            self.handle_close()
+        else:
+            try:
+                self.wfile.flush()
+            except socket.error:
+                self.handle_error()
+            else:
+                self._can_read = True
+
+    def handle_error(self):
+        logger.exception('Exception in {}!'.format(repr(self)))
+        self.handle_close()
+
+    def close(self):
+        WSGIRequestHandler.finish(self)
+        asyncore.dispatcher.close(self)
+
+    def _switch_to_websocket(self):
+        self._can_read = self._can_write = False
+        self.ws_handler_class(self.server, self.request,
+                              self.client_address, self,
+                              self._map)
+
+
+class AsyncWsgiServer(asyncore.dispatcher, WSGIServer):
+    """Asynchronous WSGI server"""
+    def __init__(self, server_address,
+                 RequestHandlerClass=AsyncWsgiHandler,
+                 map=None):
+        if map is None:
+            map = {}
+        asyncore.dispatcher.__init__(self, map=map)
+        WSGIServer.__init__(self, server_address, RequestHandlerClass, False)
+        self._poll_func = get_poll_func()
+        self.create_socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.set_reuse_addr()
+        self.bind(server_address)
+        self.listen(5)
+        self.server_address = self.socket.getsockname()
+        host, port = self.server_address[:2]
+        self.server_name = socket.getfqdn(host)
+        self.server_port = port
+        self.setup_environ()
+
+    def writable(self):
+        return False
+
+    def handle_accept(self):
+        try:
+            pair = self.accept()
+        except socket.error:
+            logger.exception('Exception when accepting a request!')
+        else:
+            if pair is not None:
+                self.RequestHandlerClass(pair[0], pair[1], self, self._map)
+
+    def handle_error(self, *args, **kwargs):
+        logger.exception('Exception in {}!'.format(repr(self)))
+        self.handle_close()
+
+    def poll_once(self, timeout=0.0):
+        """
+        Poll active sockets once
+
+        This method can be used to allow aborting server polling loop
+        on some condition.
+
+        :param timeout: polling timeout
+        """
+        if self._map:
+            self._poll_func(timeout, self._map)
+
+    def handle_request(self):
+        """Call :meth:`poll_once`"""
+        self.poll_once(0.5)
+
+    def serve_forever(self, poll_interval=0.5):
+        """
+        Start serving HTTP requests
+
+        This method blocks the current thread.
+
+        :param poll_interval: polling timeout
+        :return:
+        """
+        logger.info('Starting server on {}:{}...'.format(
+            self.server_name, self.server_port)
+        )
+        while True:
```

Comment:
Shouldn't you block with a monitor.abortRequested call?

Reply:
This is a general purpose server library. Web-PDB does not use this method. I decided to include `asyncore-wsgi` into Web-PDB addon instead of creating yet another dependency that nobody uses except me.